### PR TITLE
Cache mapped wires in `Device`

### PR DESCRIPTION
**Pull request by antalszava**
**Context:**

Wires used internally in devices get mapped based on the initial map wiring defined. Although doing this wire mapping is not too computationally intensive when done a single time, when running an optimization task where it's done thousands of times, the cost for computing mapped wires is non-negligible in terms of time.

Often devices query the mapping for the same wires several times. As the wire mapping of a device is set only when the device is created, it is safe to assume that it's immutable. Therefore we can assume that the value of mapping wires will also not change over the life of the device.

**Description of the Change:**

* When calling `map_wires`, the `Device` class now caches the key-value pair of the wires and the mapped wires in a dictionary
* The `hash` of a `Wires` object is cached

**Benefits:**
More performant device and `Wires` class usage.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
N/A

### CODE REVIEW

**Comment by antalszava** [comment]
Some benchmarks:

*GradientComputation_light.time_gradient*

Commit 8fe31ad619222e2992130c5d036571f6ef7ab51c

              --                                        n_layers / interface      
              --------- -----------------------------------------------------------------------------
               n_wires   3 / autograd    3 / tf    3 / torch    6 / autograd    6 / tf    6 / torch  
              ========= ============== ========== ============ ============== ========== ============
                  2      9.21±0.09ms    33.6±6ms   14.7±0.2ms     15.0±2ms     52.0±2ms   40.5±0.9ms 
                  5       20.0±0.1ms    92.0±3ms   78.2±0.9ms     36.5±1ms     172±8ms     328±40ms  
              ========= ============== ========== ============ ============== ========== ============


Commit 051f845aae5ed7ed6f91a0eeac0884014916aa86

              --                                        n_layers / interface                               
              --------- -----------------------------------------------------------------------------------
               n_wires   3 / autograd      3 / tf      3 / torch    6 / autograd     6 / tf     6 / torch  
              ========= ============== ============= ============= ============== ============ ============
                  2      5.13±0.03ms    16.6±0.08ms   8.26±0.04ms    7.91±0.1ms    29.8±0.1ms   22.6±0.2ms 
                  5       11.0±0.2ms      45.2±4ms      43.6±7ms      22.6±2ms      85.3±1ms     154±20ms  
              ========= ============== ============= ============= ============== ============ ============


*ML_light*

Commit 8fe31ad619222e2992130c5d036571f6ef7ab51c

Mem:

                autograd   256M 
                 torch     283M 
                   tf      303M 

Time:

                autograd   10.9±0.4s 
                 torch       failed  
                   tf      23.8±0.4s 

Commit 051f845aae5ed7ed6f91a0eeac0884014916aa86

Mem:

                autograd   256M 
                 torch     284M 
                   tf      303M 

Time:

                autograd   9.55±0.3s 
                 torch       failed  
                   tf      23.2±0.3s

File: tests/test_device.py [hunk 1]
```diff
@@ -478,6 +478,21 @@ def circuit_gauss(mag_alpha, phase_alpha, phi):
...
 
+    wires_to_try = [
+        (1, Wires([0]), Wires([0])),
+        (4, Wires([1, 3]), Wires([1, 3])),
+        (["a", 2], Wires([2]), Wires([1])),
+        (["a", 2], Wires([2, "a"]), Wires([1, 0])),
+    ]
+
+    @pytest.mark.parametrize("dev_wires, wires_to_map, res", wires_to_try)
+    def test_map_wires_caches(self, dev_wires, wires_to_map, res, mock_device):
```
Comment (mariaschuld) [inline tests/test_device.py:489]: Do you want to add a docstring?

File: tests/test_wires.py [hunk 2]
```diff
@@ -355,3 +355,9 @@ def test_equal_to_tuple(self):
         assert Wires([1, 2, 3]) == (1, 2, 3)
         assert Wires([1, 2, 3]) != (1, 5, 3)
         assert (1, 5, 3) != Wires([1, 2, 3])
+
+    def test_hash_cached(self):
+        wires = Wires([0, 1])
+        assert wires._hash is None
+        h = hash(wires)
```
Comment (mariaschuld) [inline tests/test_wires.py:362]: Same here.

File: pennylane/_device.py [hunk 3]
```diff
@@ -340,6 +341,9 @@ def map_wires(self, wires):
         Returns:
             Wires: wires with new labels
         """
+        if wires in self._cached_wires:
+            return self._cached_wires[wires]
+
```
Comment (mariaschuld) [inline pennylane/_device.py:346]: Nice solution for now. When we are refactoring the device class (if this ever happens :)  I will help turning all the current patches into a runtime-conscious solution overall!

**Review by mariaschuld — APPROVED** [review]

File: tests/test_device.py (hunk 1 above)
Comment (antalszava) [reply tests/test_device.py:489]: Good catches! Thanks, updated

**Review by josh146 — APPROVED** [review]
Looks good from my end! Agree with @mariaschuld re: having a more runtime specific solution.

A very minor thought that crossed my head while writing this - would `@functools.lru_cache` achieve the same result with less complexity?

File: pennylane/_device.py [hunk 4]
```diff
@@ -20,6 +20,7 @@
 from collections import OrderedDict, namedtuple
 
 import numpy as np
+from functools import lru_cache
```
Comment (josh146) [inline pennylane/_device.py:23]:
```suggestion
from collections import OrderedDict, namedtuple
from functools import lru_cache

import numpy as np
```

**Pull request merged**